u_prng.go

```diff
@@ -69,68 +61,36 @@ func newPRNG() (*prng, error) {
 	if err != nil {
 		return nil, err
 	}
-	return newPRNGWithSeed(seed), nil
+	return newPRNGWithSeed(seed)
 }
 
 // newPRNGWithSeed initializes a new PRNG using an existing seed.
-func newPRNGWithSeed(seed *PRNGSeed) *prng {
+func newPRNGWithSeed(seed *PRNGSeed) (*prng, error) {
+	shake := sha3.NewShake256()
+	_, err := shake.Write(seed[:])
+	if err != nil {
+		return nil, err
+	}
 	p := &prng{
-		randomStreamSeed: seed,
+		randomStream: shake,
 	}
-	p.rekey()
 	p.rand = rand.New(p)
-	return p
+	return p, nil
 }
 
 // Read reads random bytes from the PRNG stream into b. Read conforms to
 // io.Reader and always returns len(p), nil.
 func (p *prng) Read(b []byte) (int, error) {
-
 	p.randomStreamMutex.Lock()
 	defer p.randomStreamMutex.Unlock()
 
-	// Re-key before reaching the 2^38-64 chacha20 key stream limit.
-	if p.randomStreamUsed+uint64(len(b)) >= uint64(1<<38-64) {
-		p.rekey()
-	}
-
-	p.randomStream.KeyStream(b)
-
-	p.randomStreamUsed += uint64(len(b))
+	// ShakeHash.Read never returns an error:
+	// https://godoc.org/golang.org/x/crypto/sha3#ShakeHash
+	_, _ = io.ReadFull(p.randomStream, b)
 
 	return len(b), nil
```

Comment:
Why not return the error from `io.ReadFull()` anyway?
I don't think it is likely that Read() will start returning errors, but it is a package in `/x/` after all, meaning they can break things.

Reply:
That would mean propagating that error back through all the PRNG helper functions, so every `FlipWeightedCoin` etc. call would have to check errors. One of the nice properties of the original implementation was that there are no error returns on the helpers.

I think the Go convention of the documentation being a contract means it's reasonably safe to assume that this won't be changed lightly:

"// Read... It never returns an error.": https://godoc.org/golang.org/x/crypto/sha3#ShakeHash

Reply:
> That would mean propagating that error back through all the PRNG helper functions, so every FlipWeightedCoin etc. call would have to check errors.

We can zero out (by setting it to `_`) the already-nil-error in those functions, instead of ignoring it inside of `(p *prng) Read()`. In other words, uTLS can offer the same contract: `prng.Read()` always returns the nil error, but it relies on the ShakeHash.Read() contract to do that. This way, if ShakeHash's contract is ever violated, prng will at least return an error, for whatever that is worth.
Feel free to leave it as it is.